r/R/dplyr.R

```diff
@@ -276,13 +278,48 @@ source_data <- function(x) {
   }
 }
 
-is_collapsed <- function(x) inherits(x$.data, "arrow_dplyr_query")
+all_sources <- function(x) {
+  if (is.null(x)) {
+    x
+  } else if (!inherits(x, "arrow_dplyr_query")) {
+    list(x)
+  } else {
+    c(
+      all_sources(x$.data),
+      all_sources(x$join$right_data),
+      all_sources(x$union_all$right_data)
+    )
+  }
+}
 
-has_aggregation <- function(x) {
-  # TODO: update with joins (check right side data too)
-  !is.null(x$aggregations) || (is_collapsed(x) && has_aggregation(x$.data))
+query_can_stream <- function(x) {
```

Comment:
The reason we can't push this down to C++ is because we haven't constructed an exec plan yet, right? Otherwise, it would be more maintainable to do so.

Reply:
I don't follow. We could build an ExecPlan, but it wouldn't tell us anything about how it would perform, would it? I'm trying to detect cases where I can just take `head()` of the data without having to scan an entire dataset.

Reply:
> We could build an ExecPlan, but it wouldn't tell us anything about how it would perform, would it?

I'm not super close to the ExecPlan code, but I thought they were composed of a graph of nodes that could be traversed and analyzed, just like our `arrow_dplyr_query` structure. Am I wrong on that?

> I'm trying to detect cases where I can just take head() of the data without having to scan an entire dataset.

I was just thinking that having such a method on `ExecPlan` would be useful in general.

Reply:
Sure, that probably would be useful

Reply:
> I was just thinking that having such a method on ExecPlan would be useful in general.

Possibly.  We'd probably want to define it more formally.  SQL has `LIMIT X` and Substrait's equivalent is [`FetchRel`](https://substrait.io/relations/logical_relations/#fetch-operation).  Neither of these are exactly what is being detected here.  For example, it is legal to have `SELECT SUM(x) FROM table LIMIT 1` but it wouldn't actually limit any data being read.

We could define it as "single pipeline queries" but a pipeline breaker doesn't necessarily mean a query is non-streaming (for example, hash-join is sometimes permitted as "streaming" in this example but it is always a pipeline breaker).

Reply:
Since you mentioned `limit`, I'll make a plug for ARROW-16628. Not relevant for this particular question, just would let me delete some R specific handling outside of the ExecPlan, and I'm guessing we'll have to do it to support substrait.